crates/sui-package-resolver/src/lib.rs

```diff
@@ -921,52 +1066,96 @@ impl<'l> ResolutionContext<'l> {
+                        // struct) to be a runtime ID, because that's what the resolution context uses
+                        // for keys.  Take care to do this before generating the key that is used to
+                        // query and/or write into `self.structs.
+                        s.address = context.runtime_id;
+                        let key = DatatypeRef::from(s.as_ref()).as_key();
+
+                        if def.type_params.len() != s.type_params.len() {
+                            return Err(Error::TypeArityMismatch(
+                                def.type_params.len(),
+                                s.type_params.len(),
+                            ));
+                        }
 
-                    check_max_limit!(
-                        TooManyTypeParams, self.limits;
-                        max_type_argument_width >= s.type_params.len()
-                    );
+                        check_max_limit!(
+                            TooManyTypeParams, self.limits;
+                            max_type_argument_width >= s.type_params.len()
+                        );
 
-                    for (param, def) in s.type_params.iter_mut().zip(def.type_params.iter()) {
-                        if !def.is_phantom || visit_phantoms {
-                            push_ty_param!(param);
+                        for (param, def) in s.type_params.iter_mut().zip(def.type_params.iter()) {
+                            if !def.is_phantom || visit_phantoms {
+                                push_ty_param!(param);
+                            }
                         }
-                    }
 
-                    if self.structs.contains_key(&key) {
-                        continue;
-                    }
+                        if self.structs.contains_key(&key) {
+                            continue;
+                        }
 
-                    if visit_fields {
-                        for (_, sig) in &def.fields {
-                            self.add_signature(sig.clone(), store, &context, visit_fields)
-                                .await?;
+                        if visit_fields {
+                            for (_, sig) in &def.fields {
+                                self.add_signature(sig.clone(), store, &context, visit_fields)
+                                    .await?;
+                            }
                         }
-                    }
 
-                    check_max_limit!(
-                        TooManyTypeNodes, self.limits;
-                        max_type_nodes > self.structs.len()
-                    );
+                        check_max_limit!(
+                            TooManyTypeNodes, self.limits;
+                            max_type_nodes > self.structs.len()
+                        );
+
+                        self.structs.insert(key, def);
+                    } else {
```

Comment:
`T` is a `TypeTag` -- there is no `T::Enum`.

Reply:
Yea, it can either be a struct or an enum now, so it's not a great name. I really want to rename `TypeTag::Struct` to `TypeTag::Datatype`, but I'm going to hold off on that renaming till everything is in since that will just clutter a bunch of things up.